src/shared/sleep-config.c

```diff
@@ -102,6 +102,39 @@ int parse_sleep_config(SleepConfig **ret_sleep_config) {
         return 0;
 }
 
+int battery_is_low(void) {
+        _cleanup_free_ char *bat_cap_level = NULL;
+        int r;
+
+        r = read_one_line_file("/sys/class/power_supply/BAT0/capacity_level", &bat_cap_level);
+        if (r < 0)
+               return log_debug_errno(r, "Failed to read /sys/class/power_supply/BAT0/capacity_level: %m");
+
+        log_debug("Current battery level: %s", bat_cap_level);
+
+        return STR_IN_SET(bat_cap_level, "Critical", "Low");
+}
+
+int read_battery_capacity_percentage(void) {
```

Comment:
modern batteries actually expose the charge levels with much better accuracy than percentages, i.e. the "energy_full" and "energy_now" sysfs attributes. We probably should use those preferable if we can read them, and only use percentages if we don't have anything better.

hence maybe make this function return a fractional, i.e. a pair of two values, one reflecting the current level and the other the maximum level. and do not declare the unit of these two values.

Reply:
mmh but what would be the utility of doing that? Certainly from a user logging perspective a percentage is much nicer to read, having a 6 or 7 digit number dumped in front of me wouldn't help that much...
And it's not really any different for the calculations?

Reply:
percentages are good for user display, but shit for accuracy. Hence I'd alwas do calculations with the accurate units wherever we can, but when we actually want to display it, we can convert to percentages easily.

I'd also be fine if we always converted the values to 32.32 fix point values, if you want. that should provide sufficient accuracy for most purposes, but easily convert to percentages for display.

Maybe even a "double" would be OK (though I am not a fan of fp for no need), normalized to 1. 

All I am saying, doing maths based on integer percentages is ugly if we easily can get something with better accuracy.

Reply:
But it depends on the accuracy you are going for - do we need anything more precise than what we get from the percentage here? What I mean is, we really don't need a sleep timer that is accurate to the nanoseconds, once it's in the right quarter hour range then we should be good?